File: Src/Icm.Core/CommandLine/ValuesStore.vb
Imports res = My.Resources.CommandLineTools

Namespace Icm.CommandLineTools

    Public Class ValuesStore

        Private ReadOnly _store As New Dictionary(Of String, List(Of String))()
        Private ReadOnly _mainStore As New List(Of String)()

        Public Sub Clear()
            _store.Clear()
            _mainStore.Clear()
        End Sub

        Public Function ContainsKey(ByVal s As String) As Boolean
            Return _store.ContainsKey(s)
        End Function

        Public Sub AddParameter(ByVal shortName As String, ByVal longName As String)
            If shortName Is Nothing Then
                Throw New ArgumentNullException("shortName")
            End If
            If longName Is Nothing Then
                Throw New ArgumentNullException("longName")
            End If
            If shortName.Length >= longName.Length Then
                Throw New ArgumentException(String.Format(res.S_ERR_SHORT_LONGER_THAN_LONG, shortName, longName))
            End If
            _store.Add(shortName, New List(Of String))
            _store.Add(longName, New List(Of String))
        End Sub

        Public Sub RemoveValues(ByVal shortName As String, ByVal longName As String)
            _store(shortName).Clear()
            _store(longName).Clear()
        End Sub

        Public Sub AddValue(ByVal shortName As String, ByVal longName As String, ByVal val As String)
            _store(shortName).Add(val)
            _store(longName).Add(val)
        End Sub

        Public Sub AddValues(ByVal shortName As String, ByVal longName As String, ByVal vals As IEnumerable(Of String))
            _store(shortName).AddRange(vals)
            _store(longName).AddRange(vals)
        End Sub

        Public Sub AddMainValue(ByVal val As String)
            _mainStore.Add(val)
        End Sub

        Public Sub AddMainValues(ByVal vals As IEnumerable(Of String))
            _mainStore.AddRange(vals)
        End Sub

        Public ReadOnly Property Values(ByVal s As String) As ICollection(Of String)
            Get
                If _store.ContainsKey(s) Then
                    Return _store(s).AsReadOnly
                Else
                    Return (New List(Of String)).AsReadOnly
                End If
            End Get
        End Property

        Public ReadOnly Property MainValues() As ICollection(Of String)
            Get
                Return _mainStore.AsReadOnly
            End Get
        End Property
    End Class
End Namespace

File: Src/Icm.Core/CommandLine/ValuesStore.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using res = My.Resources.CommandLineTools;

namespace Icm.CommandLineTools
{

	public class ValuesStore
	{

		private readonly Dictionary<string, List<string>> _store = new Dictionary<string, List<string>>();

		private readonly List<string> _mainStore = new List<string>();
		public void Clear()
		{
			_store.Clear();
			_mainStore.Clear();
		}

		public bool ContainsKey(string s)
		{
			return _store.ContainsKey(s);
		}

		public void AddParameter(string shortName, string longName)
		{
			if (shortName == null) {
				throw new ArgumentNullException("shortName");
			}
			if (longName == null) {
				throw new ArgumentNullException("longName");
			}
			if (shortName.Length >= longName.Length) {
				throw new ArgumentException(string.Format(res.S_ERR_SHORT_LONGER_THAN_LONG, shortName, longName));
			}
			_store.Add(shortName, new List<string>());
			_store.Add(longName, new List<string>());
		}

		public void RemoveValues(string shortName, string longName)
		{
			_store(shortName).Clear();
			_store(longName).Clear();
		}

		public void AddValue(string shortName, string longName, string val)
		{
			_store(shortName).Add(val);
			_store(longName).Add(val);
		}

		public void AddValues(string shortName, string longName, IEnumerable<string> vals)
		{
			_store(shortName).AddRange(vals);
			_store(longName).AddRange(vals);
		}

		public void AddMainValue(string val)
		{
			_mainStore.Add(val);
		}

		public void AddMainValues(IEnumerable<string> vals)
		{
			_mainStore.AddRange(vals);
		}

		public ICollection<string> Values {
			get {
				if (_store.ContainsKey(s)) {
					return _store(s).AsReadOnly;
				} else {
					return (new List<string>()).AsReadOnly;
				}
			}
		}

		public ICollection<string> MainValues {
			get { return _mainStore.AsReadOnly; }
		}
	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
